pkg/daemon/update.go

```diff
@@ -1158,17 +1158,25 @@ func (dn *CoreOSDaemon) switchKernel(oldConfig, newConfig *mcfgv1.MachineConfig)
 		return nil
 	}
 
+	oldKtype := canonicalizeKernelType(oldConfig.Spec.KernelType)
+	newKtype := canonicalizeKernelType(newConfig.Spec.KernelType)
+
 	// Do nothing if both old and new KernelType are of type default
-	if canonicalizeKernelType(oldConfig.Spec.KernelType) == ctrlcommon.KernelTypeDefault && canonicalizeKernelType(newConfig.Spec.KernelType) == ctrlcommon.KernelTypeDefault {
+	if oldKtype == ctrlcommon.KernelTypeDefault && newKtype == ctrlcommon.KernelTypeDefault {
 		return nil
 	}
 
+	// TODO: Drop this code and use https://github.com/coreos/rpm-ostree/issues/2542 instead
 	defaultKernel := []string{"kernel", "kernel-core", "kernel-modules", "kernel-modules-extra"}
+	// Note this list explicitly does *not* include kernel-rt as that is a meta-package that tries to pull in a lot
+	// of other dependencies we don't want for historical reasons.
+	// kernel-rt also has a split off kernel-rt-kvm subpackage because it's in a separate subscription in RHEL.
 	realtimeKernel := []string{"kernel-rt-core", "kernel-rt-modules", "kernel-rt-modules-extra", "kernel-rt-kvm"}
 
-	dn.logSystem("Initiating switch from kernel %s to %s", canonicalizeKernelType(oldConfig.Spec.KernelType), canonicalizeKernelType(newConfig.Spec.KernelType))
+	dn.logSystem("Initiating switch from kernel %s to %s", oldKtype, newKtype)
 
-	if canonicalizeKernelType(oldConfig.Spec.KernelType) == ctrlcommon.KernelTypeRealtime && canonicalizeKernelType(newConfig.Spec.KernelType) == ctrlcommon.KernelTypeDefault {
+	switchingToThroughput := oldKtype == ctrlcommon.KernelTypeRealtime && newKtype == ctrlcommon.KernelTypeDefault
```

Comment:
question: Wouldn't it be clearer to use `switchingToDefault` instead of `switchingToThroughput` instead?

I'm a bit confused on where `throughput` comes from.

Reply:
> I'm a bit confused on where throughput comes from.

Yeah sorry that's just me having scars from years of people saying e.g. "normal RHEL" with the implication that e.g. RHEL CoreOS is not-normal.  (Or people say "normal Fedora" etc. versus "Silverblue").  Or less pejoratively they say "default RHEL"...which isn't *bad* but is also not super descriptive because, hey maybe one day what the default is changes :wink: 

From the kernel side you could certainly say `kernel` is the default.  But it is *really* about latency (kernel-rt) versus throughput (kernel).  And I personally find this is a better description.

(Also IMO the "realtime" kernel is a bit of a misnomer because it's really soft real-time which is actually quite different from [hard real time](https://en.wikipedia.org/wiki/Real-time_computing), so I personally think calling it "latency optimized" is better)

Reply:
And to expand on this, personally I'd rename `kernel` -> `kernel-throughput-optimized` and `kernel-rt` -> `kernel-latency-optimized`, I'd also rename "rhel coreos" => "rhel (image mode)" and most cruically "rhel" => "rhel (package mode)" - but only where it matters; otherwise they're both just RHEL.  Just like how both `kernel-throughput-optimized` and `kernel-latency-optimized` are both really just Linux (aka `kernel`) in different modes.

Or to say it another way, *both* are normal.  We don't strictly think of one as "default" even.  They both have qualifiers, but only where it matters.

Reply:
And to expand even more, of course today we say "OpenShift" and "Hypershift" - with the implication that the latter is the different/not-normal case.  I've even seen people refer to current OpenShift as "normal" OpenShift!  But in fact "hypershift" is (and should be!) well on its way to becoming the default, so it's also like OpenShift => OpenShift (standalone) and Hypershift => OpenShift (hosted control plane) etc.

Reply:
But all this aside, actually I was not consistent in trying to use "throughput" instead of "default" and the top patch stops using "throughput" anyways :smile:

Reply:
Thanks for the very detailed clarification! Naming things is hard and even when we think it's easy, overloaded names, changing contexts, history, etc. all make things even more complicated. I'm OK with the name `throughput` now.